mod-audit-server/src/main/java/org/folio/util/marc/ParsedRecordUtil.java

```diff
@@ -0,0 +1,243 @@
+package org.folio.util.marc;
+
+import io.vertx.core.json.JsonObject;
+import lombok.experimental.UtilityClass;
+import org.folio.dao.marc.MarcAuditEntity;
+
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Objects;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Utility class for mapping and transforming parsed MARC records into entities suitable
+ * for audit and logging purposes. This class provides methods to process and compare parsed
+ * MARC records and extract relevant data for creating audit entities.
+ * <p>
+ * The {@code ParsedRecordUtil} handles operations such as:
+ * - Mapping a source record domain event to a MarcAuditEntity.
+ * - Extracting user IDs from metadata.
+ * - Flattening complex parsed record structures into more manageable formats.
+ * - Calculating differences between old and new parsed records for auditing purposes.
+ * - Formatting and restructuring MARC field data.
+ * </p>
+ * <p>
+ * This class is intended to function as a utility and should not be instantiated directly.
+ */
+@UtilityClass
+public class ParsedRecordUtil {
+  private static final String SUBFIELDS_KEY = "subfields";
+  private static final String FIELD_KEY = "field";
+  private static final String FIELDS_KEY = "fields";
+  private static final String LEADER_KEY = "leader";
+  private static final String VALUE_KEY = "value";
+  private static final String CREATED_BY = "createdByUserId";
+  private static final String UPDATED_BY = "updatedByUserId";
+
+  /**
+   * Maps a {@link SourceRecordDomainEvent} to a {@link MarcAuditEntity}.
+   *
+   * @param event the source record domain event containing the details for creating the MarcAuditEntity;
+   *              it includes event metadata, event type, payload with new and old records, and event ID.
+   * @return a {@link MarcAuditEntity} instance containing data inferred from the given event, such as event ID,
+   * event date, record ID, origin, action, user ID, record type, and any differences.
+   */
+  public static MarcAuditEntity mapToEntity(SourceRecordDomainEvent event) {
+    String userId;
+    String recordId;
+    SourceRecordType recordType;
+    Map<String, Object> difference;
+    var oldRecord = event.getEventPayload().getOld();
+    var newRecord = event.getEventPayload().getNewRecord();
+
+    switch (event.getEventType()) {
+      case SOURCE_RECORD_CREATED -> {
+        recordId = newRecord.getId();
+        userId = getUserIdFromMetadata(newRecord.getMetadata(), CREATED_BY);
+        recordType = newRecord.getRecordType();
+        difference = getDifference(newRecord.getParsedRecord(), event.getEventType());
```

Comment:
@dmytrokrutii  Is there an opportunity to extract it to the separate method, and provide different arguments, to avoid code duplication in 3 cases? If not - feel free to resolve this comment.